Single-File: Process bundles in the framework

swaroop-sridhar opened the pull request:
This change implements the host changes proposed in the [design](https://github.com/dotnet/designs/blob/master/accepted/2020/single-file/design.md#startup)

The main changes for single-file bundles are:
* Bundle processing code is moved from apphost to hostpolicy
* HostFxr and HostPolicy process deps.json and runtimeconfig.json files directly from the bundle.
* HostPolicy performs verification wrt deps.json based on the contents of the single-file bundle.
* AppContext.BaseDirectory is set as explained [here](https://github.com/dotnet/designs/blob/master/accepted/2020/single-file/design.md#appcontextbasedirectory)

Currently, all files except deps.json and runtimeconfig.json are extracted to disk.
Once the runtime is able to processing assemblies directly from the bundle, they will no longer be extracted.

Notable details:
* The bundle driver (formarly runner.cpp) is divided into two parts:
   * bundle::info which describes basic information about the bundle available from the headers only
   * bundle::runner which has information about all embedded files, and the ability to drive extraction
  This facilitates linking only parts of the bundle handling code with hostfxr, while all code is linked with hostpolicy.
* The AppHost only links with bundle_marker to identify itself as a single-file bundle.
* If the AppHost is a single-file bundle, it notifies hostfxr using the new hostfxr_main_bundle_startup_info() API
* The HostFxr comminucates the single-file-information with HostPolicy using the host_interface_t structure.

Fixes https://github.com/dotnet/runtime/issues/32821

swaroop-sridhar committed:
Single-File: Process bundles in the framework

This change implements the host changes proposed in the [design](https://github.com/dotnet/designs/blob/master/accepted/2020/single-file/design.md#startup)

The main changes for single-file bundles are:
* Bundle processing code is moved from apphost to hostpolicy
* HostFxr and HostPolicy process deps.json and runtimeconfig.json files directly from the bundle.
* HostPolicy performs verification wrt deps.json based on the contents of the single-file bundle.
* AppContext.BaseDirectory is set as explained [here](https://github.com/dotnet/designs/blob/master/accepted/2020/single-file/design.md#appcontextbasedirectory)

Currently, all files except deps.json and runtimeconfig.json are extracted to disk.
Once the runtime is able to processing assemblies directly from the bundle, they will no longer be extracted.

Notable details:
* The bundle driver (formarly runner.cpp) is divided into two parts:
   * bundle::info which describes basic information about the bundle available from the headers only
   * bundle::runner which has information about all embedded files, and the ability to drive extraction
  This facilitates linking only parts of the bundle handling code with hostfxr, while all code is linked with hostpolicy.
* The AppHost only links with bundle_marker to identify itself as a single-file bundle.
* If the AppHost is a single-file bundle, it notifies hostfxr using the new hostfxr_main_bundle_startup_info() API
* The HostFxr comminucates the single-file-information with HostPolicy using the host_interface_t structure.

Fixes https://github.com/dotnet/runtime/issues/32821

swaroop-sridhar committed:
Fix tests

swaroop-sridhar committed:
Tests: Publish files excluded from bundle to bundle directory

swaroop-sridhar committed:
Fix a few test leg failures

lpereira commented on src/installer/corehost/cli/json/rapidjson/cursorstreamwrapper.h:
```diff
@@ -46,7 +46,7 @@ class CursorStreamWrapper : public GenericStreamWrapper<InputStream, Encoding> {
     Ch Take() {
         Ch ch = this->is_.Take();
-        if(ch == '\n') {
+        if (ch == '\n') {
```
This is a harmless change, but I'd prefer if rapidjson were kept as a pristine copy from upstream, as it's a 3rd-party component.  (If we need to actually fix things in there, I'd prefer that we also sent patches their way instead of maintaining a fork.)

swaroop-sridhar replied:
OK @lpereira  I'll back off the change from this PR.
I've submitted a PR for rapidjson: https://github.com/Tencent/rapidjson/pull/1685
We should subsequently fix it in our code too. 
I usually do `if(`  `for(` `while(` searches before submitting PRs, and it would be nice to not hit these cases every time.

lpereira commented on src/installer/corehost/cli/json_parser.cpp:
```diff
@@ -130,5 +135,41 @@ bool json_parser_t::parse_stream(pal::istream_t& stream,
+#else //  _WIN32
+            // On Linux, copy the input data, since m_document.ParseInsitu
+            // requires a mutable data input.
+            realloc_buffer(location->size);
+            memcpy(m_json.data(), data, location->size);
```
I usually see the idiom of using the `swap()` method to initialize a `std::vector` from an array, so I was curious to see how Clang would generate the code. [It's identical](https://gcc.godbolt.org/z/9P7Qn_). I'm impressed. :)

lpereira commented on src/installer/corehost/cli/bundle/info.cpp:
```diff
@@ -0,0 +1,145 @@
+    // and return the address at the appropriate offset.
+    // This is because:
+    // * The host is the only code that is currently running, and attempting to map the bundle at this time
+    // * Files can only be memory mapped at page-aligned offsets, and in whole page units.
+    //   Therefore, mapping only portions of the bundle will involve align-down/round-up calculations, and associated offset adjustments.
+    //   We choose the simpler approach of rounding to the whole file
```
Might be a good idea to comment that this might be a problem in 32-bit systems and very large applications due to exhaustion of address space.

lpereira commented on src/installer/corehost/cli/bundle/info.h:
```diff
@@ -0,0 +1,100 @@
+        private:
+            pal::string_t m_path;
+            const location_t *m_location;
+        } json_info;
```
`json_info` doesn't seem to be used from a quick glance.  Might want to remove the `config_t()` constructor, then, because method such as `matches()` will reference a garbage pointer

lpereira commented on src/installer/corehost/cli/bundle/info.h:
```diff
@@ -0,0 +1,100 @@
+    {
+        struct config_t
+        {
+            config_t() {}
```
This doesn't initialize `m_location`, and `config_t::matches()` will dereference that pointer.  When `info_t::m_deps_json`and `info_t::m_runtimeconfig_json` are constructed, they use this constructor here.

lpereira commented on src/installer/corehost/cli/bundle/manifest.cpp:
```diff
@@ -12,7 +12,9 @@ manifest_t manifest_t::read(reader_t& reader, int32_t num_files)
     {
-        manifest.files.emplace_back(file_entry_t::read(reader));
+        file_entry_t entry = file_entry_t::read(reader);
+        manifest.files.emplace_back(entry);
```
Nit: Is there a benefit for using `emplace_back()`here? Maybe `file_entry_t` should have a move constructor and `std::move`be used instead?

lpereira commented on src/installer/corehost/cli/bundle/info.cpp:
```diff
@@ -0,0 +1,145 @@
+const int8_t* info_t::config_t::map(const pal::string_t& path, const location_t* &location)
+{
+    const bundle::info_t* app = bundle::info_t::the_app;
+
```
Since `the_app` is static and initialized to `nullptr`, might be a good idea to assert if it's not `nullptr` where it's dereferenced

lpereira commented on src/installer/corehost/cli/bundle/runner.cpp:
```diff
@@ -0,0 +1,79 @@
+    pal::string_t file_base = entry->needs_extraction() ? app->extraction_path() : app->base_path();
+
+    // Reserve space for the path below
+    full_path.reserve(file_base.length() + relative_path.length() + 3);
```
What's the `3` for?

swaroop-sridhar replied:
This is code copied from https://github.com/dotnet/runtime/blob/master/src/installer/corehost/cli/deps_entry.cpp#L32.
Two characters are required for `DIR_SEPARATOR` and null-terminator. Not sure why the third spot is necessary.

lpereira commented on src/installer/corehost/cli/deps_entry.cpp:
```diff
@@ -6,7 +6,45 @@
+{
+    assert(asset_type == asset_types::resources);
+
+    pal::string_t pal_relative_path = asset.relative_path;
+    if (_X('/') != DIR_SEPARATOR)
+    {
+        replace_char(&pal_relative_path, _X('/'), DIR_SEPARATOR);
+    }
```
`normalize_dir_separator()`?
```suggestion
    pal::string_t pal_relative_path = normalize_dir_separator(asset.relative_path);
```

lpereira commented on src/installer/corehost/cli/hostmisc/pal.unix.cpp:
```diff
@@ -74,11 +74,16 @@ void* pal::map_file_readonly(const pal::string_t& path, size_t& length)
         close(fd);
         return nullptr;
     }
+    size_t size = buf.st_size;
```
`size_t` is only for size in memory; for files, you want `off_t`.

swaroop-sridhar replied:
I agree that `buf.st_size` is an `off_t` but we're computing `size` to be an input for `mmap()`, which expects a `size_t length` argument .

lpereira commented on src/installer/corehost/cli/hostmisc/pal.unix.cpp:
```diff
@@ -74,11 +74,16 @@ void* pal::map_file_readonly(const pal::string_t& path, size_t& length)
+    void* address = mmap(nullptr, size, PROT_READ, MAP_SHARED, fd, 0);
 
-    if(address == nullptr)
+    if (address == nullptr)
```
```suggestion
    if (address == MAP_FAILED)
```

lpereira commented on src/installer/corehost/cli/hostmisc/pal.unix.cpp:
```diff
@@ -74,11 +74,16 @@ void* pal::map_file_readonly(const pal::string_t& path, size_t& length)
+    if (address == nullptr)
     {
         trace::error(_X("Failed to map file. mmap(%s) failed with error %d"), path.c_str(), errno);
         close(fd);
```
```suggestion
```

lpereira commented on src/installer/corehost/cli/hostpolicy/hostpolicy_init.cpp:
```diff
@@ -126,6 +127,15 @@ bool hostpolicy_init_t::init(host_interface_t* input, hostpolicy_init_t* init)
+    {
+        if (input->app_bundle != nullptr)
+        {
+            static bundle::runner_t bundle_runner(input->app_bundle);
```
I keep reading this as Blade Runner. :)

swaroop-sridhar commented on src/installer/corehost/cli/json/rapidjson/document.h:
```diff
@@ -2094,11 +2094,11 @@ class GenericValue {
         const SizeType len1 = GetStringLength();
         const SizeType len2 = rhs.GetStringLength();
-        if(len1 != len2) { return false; }
+        if (len1 != len2) { return false; }
```
https://github.com/dotnet/runtime/pull/34274/files/51c0c31d0a9cb26a6e3119bf6609cd300c90c96c#r402006698

swaroop-sridhar committed:
Address feedback from @lpereira

swaroop-sridhar committed:
Undo temporary test changes

swaroop-sridhar committed:
Fix a rename bug

swaroop-sridhar committed:
Improve a test

The rename test would hang forever if AppWithWait() fails before creating the resume file.
So, while waiting, check that the app-process is still running.

vitek-karas commented on src/installer/corehost/cli/bundle/info.cpp:
```diff
@@ -0,0 +1,147 @@
+    // When necessary to map the deps.json or runtimeconfig.json files, we map the whole single-file bundle,
+    // and return the address at the appropriate offset.
+    // This is because:
+    // * The host is the only code that is currently running, and attempting to map the bundle at this time
```
This feels like not a finished sentence...

vitek-karas commented on src/installer/corehost/cli/bundle/info.cpp:
```diff
@@ -0,0 +1,147 @@
+    // * There is no performance limitation due to a larger sized mapping, since we actually only read the pages with relevant contents.
+    // * Files that are too large to be mapped (ex: that exhaust 32-bit virtual address space) are not supported. 
+
+    int8_t* addr = (int8_t*)pal::mmap_read(app->m_bundle_path);
```
I know it's not part of this change yet... do we plan to "cache" the mapping? Once we start loading assemblies from memory I would expect to map the bundle once and hand out pointers from it as required...

swaroop-sridhar replied:
There's no plan to cache the mapping. 
* Closing the mapping enables renaming the app while it is running on Windows -- an idiom used to update apps in-place.
* The PE loader itself maps each section of the bundled assemblies at aligned addresses.

The only item that is cached is the bundle-manifest, so that bundle-probes from the runtime can be answered without need for re-processing the bundle headers.

vitek-karas replied:
Good point - I forgot that runtime will create its own mapping anyway.

vitek-karas commented on src/installer/corehost/corehost.cpp:
```diff
@@ -200,23 +187,32 @@ int exe_start(const int argc, const pal::char_t* argv[])
+        hostfxr_main_bundle_startup_info_fn main_fn_v3 = reinterpret_cast<hostfxr_main_bundle_startup_info_fn>(pal::get_symbol(fxr, "hostfxr_main_bundle_startup_info"));
+        if (main_fn_v3 == nullptr)
+        {
+            trace::error(_X("The required library %s does not support single-file apps."), fxr_path.c_str());
```
What is the experience around this:
* I take a new (5.0) single-file (FDD) app and try to run it on a machine which doesn't have 5.0 installed.
* It fails with this error which doesn't really tell me anything. In addition, for GUI apps the reported `rc` is not recognized as special and there won't be any dialog.

Ideally we want experience similar to hostfxr missing or framework missing. I should get a URL to go to, and for GUI I should get a dialog.
Basically the problem is that this is effectively a version check, which happens before we run FX resolution. So we need it to behave in a very similar manner.

swaroop-sridhar replied:
I've fixed this, here's a sample output:
```
winhello.exe
The required library C:\Program Files\dotnet\host\fxr\5.0.0-preview.3.20169.1\hostfxr.dll does not support single-file apps.
  _ To run this application, you need to install a newer version of .NET Core.
  - https://aka.ms/dotnet-core-applaunch?missing_runtime=true&arch=x64&rid=win10-x64
```
![image](https://user-images.githubusercontent.com/8977622/78415750-c5caa800-75d8-11ea-83f5-87cefeaf3080.png)

vitek-karas replied:
Can you please double-check that the URL which the dialog leads to contains apphost version information?

swaroop-sridhar replied:
Yes the link generated now is:
https://aka.ms/dotnet-core-applaunch?missing_runtime=true&arch=x64&rid=win10-x64&apphost_version=5.0.0-dev

vitek-karas commented on src/installer/corehost/corehost.cpp:
```diff
@@ -200,23 +187,32 @@ int exe_start(const int argc, const pal::char_t* argv[])
+        int64_t bundle_header_offset = bundle_marker_t::header_offset();
 
-        trace::info(_X("Invoking fx resolver [%s] v2"), fxr_path.c_str());
+        trace::info(_X("Invoking fx resolver [%s] v3"), fxr_path.c_str());
```
What is the "v3" for? It doesn't really match any version - and the entry points are not versioned in their name. Can we please remove this notion - it's just confusing.
Let's just trace the name of the entry point we're calling. (same below).
Could you also please rename the local variables to not refer to versions...

vitek-karas commented on src/installer/corehost/corehost.cpp:
```diff
@@ -200,23 +187,32 @@ int exe_start(const int argc, const pal::char_t* argv[])
-            rc = main_fn_v2(argc, argv, host_path_cstr, dotnet_root_cstr, app_path_cstr);
+            rc = main_fn_v3(argc, argv, host_path_cstr, dotnet_root_cstr, app_path_cstr,bundle_header_offset);
 
             if (trace::get_error_writer() != nullptr && rc == static_cast<int>(StatusCode::FrameworkMissingFailure) && !set_error_writer_fn)
```
This is not necessary due to the above "version" check. This was introduced here: https://github.com/dotnet/runtime/pull/1222
The reason was to get a good error message for UI apps when running 3.0 apps on 2.0 only hostfxr which doesn't support error writer redirection. In this case the failure will happen earlier and if we get here it will be with >=5.0 hostfxr which will support the fully featured error reporting.

Maybe add comments describing this reasoning here and to the case below for 3.0 app on 2.0 hostfxr.

vitek-karas reviewed (changes requested):
I have yet to look at tests. The most important feedback is about the UX when we run 5.0 single-file on machine which doesn't have 5.0. We need to get that right.

swaroop-sridhar committed:
Address feedback from vitek-karas

* Fix HostFxr startup function naming in CoreHost
* deps_resolver: Lookup up bundle even for rid-specific assets within the app
* Ensure CUI/GUI error output when expected framework is not found for single file apps.
* Other minor fixes

swaroop-sridhar committed:
Fix tests failures

swaroop-sridhar committed:
Fix a non-apphost scenario

vitek-karas commented on src/installer/corehost/corehost.cpp:
```diff
@@ -84,6 +83,14 @@ bool is_exe_enabled_for_execution(pal::string_t* app_dll)
+    pal::string_t url = get_download_url();
+    trace::error(_X("  _ To run this application, you need to install a newer version of .NET Core."));
+    trace::error(_X(""));
+    trace::error(_X("  - %s"), url.c_str());
```
Can you please append the apphost version to the URL? The problem is that without it there's no version information in the URL and the web server will have to way of telling at least roughly which version to offer. Once we ship 6.0 that would probably become the default and 5.0 apps would not work with it (default roll forward is Minor only). It would require two installations (6.0 first - which would provide better error reporting as we would load hostfxr and then 5.0 to actually run the app).

See this change which improved this for the other cases: https://github.com/dotnet/runtime/pull/33602

swaroop-sridhar committed:
Fix a bug in deps resolution

swaroop-sridhar committed:
Fix resource path

vitek-karas commented on src/installer/test/Assets/TestProjects/AppWithSubDirs/Program.cs:
```diff
@@ -3,6 +3,7 @@
 // See the LICENSE file in the project root for more information.
 
 using System;
+using System.Diagnostics;
```
Nit: probably not needed

vitek-karas reviewed (commented):
I think I 'm struggling to understand what is the hostpolicy/coreclr interface. In the design doc https://github.com/dotnet/designs/blob/master/accepted/2020/single-file/design.md#dependency-resolution it says:
>The assemblies bundled within the single-file are not enumerated in TRUSTED_PLATFORM_ASSEMBLIES. The absolute-path of assemblies on disk are listed in TRUSTED_PLATFORM_ASSEMBLIES as usual.

But if I read the code correctly it will populate TPA with full paths even for bundled assemblies - relative to the app's base.

Similar comment on resource paths.

swaroop-sridhar commented on src/installer/corehost/cli/deps_entry.cpp:
```diff
@@ -21,28 +21,29 @@ static pal::string_t normalize_dir_separator(const pal::string_t& path)
+//
+// If this file exists within the single-file bundle candidate is:
+//  * If the file was extracted to disk, the full-path to the extracted file.
+//  * Otherwise, the path within the bundle, relative to the "base" directory.
+//    The runtime expects the entries in the TPA,  NativeDllSearchDirectories, etc 
+//    to be absolute paths. Therefore, the relative-paths within the bundle are
+//    expressed as absolute paths with respect to the location of the bundle-file.
```
But I agree that this comment is confusing, because it tries to talk about both the current and future state of the machine. I'll fix it to only say that the assembly is expected to be found at the extracted location.

vitek-karas replied:
So if I read it correctly, this is a point-in-time behavior.
Currently it is expected to actually extract all assemblies on disk. And so we calculate the path as relative to the app-base.
In the future, these will not be extracted and thus the code will change to not include them in TPA (and the logic how they're resolved will have to adapt as well). Right?

swaroop-sridhar replied:
Yes, this is correct.

swaroop-sridhar committed:
Add apphost version to framework_missing URL

Also update a few comments to clarify that assemblies are currently extracted to disk.

elinor-fung commented on src/installer/corehost/cli/host_interface.h:
```diff
@@ -58,6 +59,7 @@ struct host_interface_t
     const pal::char_t* host_info_host_path;
     const pal::char_t* host_info_dotnet_root;
     const pal::char_t* host_info_app_path;
+    const bundle::info_t *app_bundle;
```
`bundle::info_t` does not look to be plain data, so using it cross-binary like this where hostpolicy reads the class allocated by hostfxr could lead to compat issues.

swaroop-sridhar replied:
Thanks @elinor-fung . I've fixed this in https://github.com/dotnet/runtime/pull/34274/commits/12b63bae12ab435120b8a9b56165a86d6563406f

elinor-fung commented on src/installer/corehost/cli/fxr/hostpolicy_resolver.cpp:
```diff
@@ -12,6 +12,7 @@
 #include <utils.h>
 
 #include "json_parser.h"
+#include "bundle/info.h"
```
Not needed?

swaroop-sridhar committed:
Fix a bug in hostpolicy initialization for nativehost

swaroop-sridhar committed:
HostInterface: Don't pass host::info pointer across hostfxr/hostpolicy.

Host::info* is not POD, so pass the bundle-header offset instead.
The deps-json and runtime-config json locations are recomputed in hostpolicy.
Since hostpolicy needs to re-map the bundle, recomputing these locations is not expensive than passing them through the interface.
This also keeps the host_interface cleaner.

vitek-karas reviewed (approved):
Looks good - the one pending question I have is the RID specific asset handling.

swaroop-sridhar committed:
Fix an mmap() call

swaroop-sridhar committed:
Fix json_parser: Unmap only during destruction.

swaroop-sridhar commented on src/installer/corehost/cli/hostpolicy/deps_resolver.cpp:
```diff
@@ -343,6 +343,12 @@ bool deps_resolver_t::probe_deps_entry(const deps_entry_t& entry, const pal::str
                 else
                 {
                     // Non-rid assets, lookup in the published dir.
```
@vitek-karas, regarding rid-specific assets: 
* I've added the bundle-check in the rid-specific assets case too, as you've recommended.
* However, when an app is published specific to a RID, my understanding is that the SDK will never create assets for other RIDs. The SDK will find the best compatible RID assets and copy them to the publish directory.  So, there won't be a `<rid>\publish\runtimes\<rid>` structure.

For example, if we have a package (libuv 1.9.0 is a very good example)
```
runtimes
   |- debian-x64/native/lib.so
   |- rhel-x64/native/lib.so
   |- win-x64/native/lib.dll
```
`dotnet publish` will generate the above directory structure within the `bin/.net5/publish` directory.

However, `dotnet publish -r debian-x64 --self-contained=false`  will generate `bin/.net5/debian-x64/publish` that *only* contains `lib.so` from `runtime/debian-x64/native`.

`dotnet publish -r ubuntu-x64 --self-contained=false`  will generate `bin/.net5/ubuntu-x64/publish` that contains `lib.so` from `runtime/debian-x64/native`.

`dotnet publish -r linux-x64 --self-contained=false`  will generate `bin/.net5/linux-x64/publish` without `lib.so`

vitek-karas replied:
> dotnet publish -r linux-x64 --self-contained=false will generate bin/.net5/linux-x64/publish without lib.so

This sounds like a problem/bug. I assume the app would not run in that case at all... right?

swaroop-sridhar replied:
I agree that there is an underlying problem, because the app may fail to run because of missing assemblies in this situation. However, I'm not sure it is an SDK bug, but rather a product issue. 

The missing RID-asset (here `linux-x64`) in a package may actually mean that the library (`lib.so`) is unnecessary for that platform -- although it is weird that the lib is not necessary at a higher RID but necessary in a lower RID. This is somewhat similar to missing an assembly that is only loaded at runtime for reflection -- where the SDK cannot know the exact requirements. 

@dsplaisted  @wli3: Should we consider guards in the SDK to check for this case? (ex: missing assets at a parent RID)? Thanks.

dsplaisted replied:
@swaroop-sridhar I have a feeling trying to add guards for this type of thing would be more complex and complicated than would be warranted for the benefit we would get out of it.  But you can file an issue in the SDK repo if you want to discuss it further.

swaroop-sridhar replied:
https://github.com/dotnet/sdk/issues/11206

swaroop-sridhar commented:
> the one pending question I have is the RID specific asset handling.

I've now answered this in https://github.com/dotnet/runtime/pull/34274#discussion_r405265681

vitek-karas reviewed (approved):
Looks good.

Please followup on the issue with RID-specific assets. I do agree that vast majority of the apps won't have the problem. If SDK will never add a RID-specific asset to a RID-specific build, then this is not an issue for host/runtime (yet), but it seems to be an SDK issue.
